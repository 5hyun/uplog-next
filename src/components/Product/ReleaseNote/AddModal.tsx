import React from 'react';
import { useMessage } from '@/hooks/useMessage.ts';
import {
  Modal,
  ModalBody,
  ModalCloseButton,
  ModalContent,
  ModalFooter,
  ModalHeader,
  ModalOverlay,
} from '@chakra-ui/react';

interface Props {
  isOpen: boolean;
  onClose: () => void;
}
export default function AddModal({ isOpen, onClose }: Props) {
  const { showMessage, contextHolder } = useMessage();

  return (
    <Modal isCentered onClose={onClose} isOpen={isOpen}>
      {contextHolder}
      <ModalOverlay />

      <ModalContent
        maxW={'41rem'}
        h={'22rem'}
        shadow={'boxShadow-sign-up'}
        rounded={'none'}
        p={'1.2rem'}
        bg={'var(--white)'}
      >
        <ModalHeader
          fontSize={'1.8rem'}
          fontWeight={700}
          bg={'var(--white)'}
          color={'var(--black)'}
        >
          프로젝트 추가
        </ModalHeader>
        <ModalCloseButton
          fontSize={'1rem'}
          color={'var(--gray-light)'}
          mt={'0.6rem'}
          mr={'0.8rem'}
        />
        <ModalBody>
          <section className={'flex flex-col justify-center items-start w-h-full'}>
            <div className={'mx-auto'}>
              <span className={'text-gray-dark font-bold text-[1.2rem]'}>버전 입력</span>
              <input
                className={
                  'border-base mt-4 mb-2 w-[29rem] h-14 rounded-xl px-4 py-2 text-[1.1rem] text-black'
                }
                type="text"
                placeholder={'새로 생성할 버전을 입력해 주세요.'}
                maxLength={10}
              />
              <span className={'text-gray-light text-[0.6rem] ml-2'}>
                * 현재 생성되는 버전은 임시 버전이며, 완료 시 최종 버전명을 재설정할 수 있습니다.
              </span>
            </div>
          </section>
        </ModalBody>

        <ModalFooter>
          <button
            className={'bg-orange rounded font-bold text-xs text-white w-[4.5rem] h-9'}
            onClick={() => {}}
          >
            완료
          </button>
        </ModalFooter>
      </ModalContent>
    </Modal>
  );
}
