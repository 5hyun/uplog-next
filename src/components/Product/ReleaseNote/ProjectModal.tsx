import React, { useCallback, useEffect, useRef } from 'react';
import {
  Modal,
  ModalBody,
  ModalCloseButton,
  ModalContent,
  ModalFooter,
  ModalHeader,
  ModalOverlay,
  Tooltip,
} from '@chakra-ui/react';
import useInput from '@/hooks/useInput.ts';
import { BsQuestionCircle } from 'react-icons/bs';
import CompleteModalTooltip from '@/components/Member/MyPage/CompleteModalTooltip.tsx';
import { completeProject, createProject } from '@/api/Project/Version.ts';
import { useMutation, useQueryClient } from 'react-query';
import { ProductInfo } from '@/typings/product.ts';
import { eachProductProjects } from '@/recoil/Project/atom.ts';
import { useRecoilState } from 'recoil';
import { Release } from '@/typings/project.ts';
type MessageType = 'success' | 'error' | 'warning';

interface Props {
  isOpen: boolean;
  onClose: () => void;
  isAdd: boolean;
  versionName: string;
  showMessage: (type: MessageType, content: string) => void;
  nowProjectId: number;
}
export default function ProjectModal({
  isOpen,
  onClose,
  isAdd,
  versionName,
  showMessage,
  nowProjectId,
}: Props) {
  const nowProduct: ProductInfo = JSON.parse(sessionStorage.getItem('nowProduct')!);
  const queryClient = useQueryClient();

  // 프로젝트 추가에서는 프로젝트 이름, 완료에서는 최종 버전
  const [text, onChangeText, setText] = useInput('');
  const [projects, setProjects] = useRecoilState(eachProductProjects);
  const newProjectId = useRef(-1);

  const { mutate: createProjectMutate } = useMutation(createProject, {
    onSuccess: (data) => {
      if (typeof data !== 'string') {
        newProjectId.current = data.id;
      }
    },
    onMutate: async ({ version }) => {
      await queryClient.cancelQueries(['getAllProductProjects', nowProduct?.productId]);

      const snapshot = queryClient.getQueryData(['getAllProductProjects', nowProduct?.productId]);

      queryClient.setQueriesData(['getAllProductProjects', nowProduct?.productId], () => {
        const temp: Release[] = [...projects, { version, projectStatus: 'PROGRESS_IN', id: -1 }];

        return temp;
      });

      return { snapshot };
    },
    onError: (error, newTodo, context) => {
      queryClient.setQueriesData(
        ['getAllProductProjects', nowProduct?.productId],
        context?.snapshot
      );
    },
    onSettled: () => {
      queryClient.invalidateQueries({ queryKey: ['getAllProductProjects', nowProduct?.productId] });
    },
  });

  const { mutate: completeProjectMutate } = useMutation(completeProject, {
    onMutate: async ({ version, projectId }) => {
      await queryClient.cancelQueries(['getAllProductProjects', nowProduct?.productId]);

      const snapshot = queryClient.getQueryData(['getAllProductProjects', nowProduct?.productId]);

      queryClient.setQueriesData(['getAllProductProjects', nowProduct?.productId], () => {
        const temp: Release[] = projects.map((project) => {
          if (project.id === projectId) {
            return { ...project, version, projectStatus: 'PROGRESS_COMPLETE' };
          } else return project;
        });

        setProjects(temp);
        return temp;
      });

      return { snapshot };
    },
    onError: (error, newTodo, context) => {
      queryClient.setQueriesData(
        ['getAllProductProjects', nowProduct?.productId],
        context?.snapshot
      );
    },
    onSettled: () => {
      queryClient.invalidateQueries({ queryKey: ['getAllProductProjects', nowProduct?.productId] });
    },
  });

  const onClickCreateProject = useCallback(() => {
    if (projects.some((project) => project.version === text)) {
      showMessage('error', '이미 존재하는 버전입니다!');
      return;
    }

    //TODO: url 수정 필요
    createProjectMutate({ productId: nowProduct?.productId, version: `${text}(임시)`, link: '/' });
    showMessage('success', '프로젝트가 생성되었습니다!');
    onClose();
    setText('');
  }, [text]);

  const onClickCompleteProject = useCallback(() => {
    if (projects.some((project) => project.version === text)) {
      showMessage('error', '이미 존재하는 버전입니다!');
      return;
    }

    completeProjectMutate({
      projectId: nowProjectId === -1 ? newProjectId.current : nowProjectId,
      version: text,
    });
    showMessage('success', '프로젝트 완료!');
    onClose();
    setText('');
  }, [text, newProjectId]);

  useEffect(() => {
    if (versionName && !isAdd) {
      setText(versionName.replace('(임시)', ''));
    } else setText('');
  }, [isAdd]);

  return (
    <Modal isCentered onClose={onClose} isOpen={isOpen}>
      <ModalOverlay />

      <ModalContent
        maxW={'41rem'}
        h={'22rem'}
        shadow={'boxShadow-sign-up'}
        rounded={'none'}
        p={'1.2rem'}
        bg={'var(--white)'}
      >
        <ModalHeader
          fontSize={'1.8rem'}
          fontWeight={700}
          bg={'var(--white)'}
          color={'var(--black)'}
        >
          {isAdd ? '프로젝트 추가' : '프로젝트 완료'}
        </ModalHeader>
        <ModalCloseButton
          fontSize={'1rem'}
          color={'var(--gray-light)'}
          mt={'0.6rem'}
          mr={'0.8rem'}
        />
        <ModalBody>
          <section className={'flex flex-col justify-center items-start w-h-full'}>
            <div className={'mx-auto'}>
              <span className={'text-gray-dark font-bold text-[1.2rem]'}>
                {isAdd ? '버전 입력' : '최종 버전 입력'}
              </span>
              <div className={'flex items-center'}>
                <input
                  className={`border-base mt-4 mb-2 ${
                    isAdd ? 'w-[29rem]' : 'w-[25rem]'
                  } h-14 rounded-xl px-4 py-2 text-[1.1rem] text-black `}
                  type="text"
                  value={text}
                  onChange={onChangeText}
                  placeholder={
                    isAdd ? '새로 생성할 버전을 입력해 주세요.' : '최종 버전을 입력해 주세요.'
                  }
                  maxLength={10}
                />

                {!isAdd && (
                  <Tooltip
                    placement={'bottom-start'}
                    label={<CompleteModalTooltip />}
                    color={'var(--black)'}
                    backgroundColor={'var(--white)'}
                    minW={'29rem'}
                    boxShadow={
                      '2px 2px 5px 1px rgba(0, 0, 0, 0.05), -2px -2px 5px 0px rgba(0, 0, 0, 0.05)'
                    }
                    borderRadius={'0.31rem'}
                  >
                    <span className={'flex-col-center mt-3 ml-4 text-gray-dark cursor-pointer '}>
                      <BsQuestionCircle className={'text-3xl mb-1'} />
                      버전
                    </span>
                  </Tooltip>
                )}
              </div>
              {isAdd && (
                <span className={'text-gray-light text-[0.6rem] ml-2'}>
                  * 현재 생성되는 버전은 임시 버전이며, 완료 시 최종 버전명을 재설정할 수 있습니다.
                </span>
              )}
            </div>
          </section>
        </ModalBody>

        <ModalFooter>
          <button
            className={'bg-orange rounded font-bold text-sm text-white w-[4.5rem] h-9'}
            onClick={() => {
              isAdd ? onClickCreateProject() : onClickCompleteProject();
            }}
          >
            완료
          </button>
        </ModalFooter>
      </ModalContent>
    </Modal>
  );
}
