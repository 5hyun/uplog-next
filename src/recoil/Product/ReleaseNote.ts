import { atom } from 'recoil';
import { changeType } from '@/typings/product.ts';

export const typeBgColors = atom<Record<changeType, string>>({
  key: 'typeBgColors',
  default: {
    FEATURE: 'bg-type-feature',
    CHANGED: 'bg-type-changed',
    FIXED: 'bg-type-fixed',
    NEW: 'bg-type-new',
    DEPRECATED: 'bg-type-deprecated',
  },
});
