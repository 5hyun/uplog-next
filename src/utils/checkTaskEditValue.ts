import { TaskData } from '@/typings/task.ts';

export function checkTaskEditValue(editTask: TaskData): boolean {
  if (editTask.taskName === '') {
    return false;
  }
  if (editTask.targetMemberInfoDTO && editTask.targetMemberInfoDTO.id === 0) {
    return false;
  }
  if (editTask.menuId === 0) {
    return false;
  }
  if (editTask.teamId === 0) {
    return false;
  }
  if (editTask.startTime === '' || editTask.endTime === '') {
    return false;
  }

  return true;
}
