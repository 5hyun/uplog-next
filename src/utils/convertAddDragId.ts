import { StatusTaskData, Tasks } from '@/typings/task.ts';
import { useState } from 'react';

export function convertAddDragId(tasks: StatusTaskData): Tasks {
  const [convertedTasks, setConvertedTasks] = useState<Tasks>({
    PROGRESS_BEFORE: [],
    PROGRESS_IN: [],
    PROGRESS_COMPLETE: [],
  });

  const beforeStatusTasks = tasks.PROGRESS_BEFORE.map((task) => ({
    ...task,
    dragId: task.id.toString(), // dragId 값을 문자열로 변환하여 할당
  }));

  setConvertedTasks({
    ...convertedTasks,
    PROGRESS_BEFORE: beforeStatusTasks,
  });

  const inStatusTasks = tasks.PROGRESS_IN.map((task) => ({
    ...task,
    dragId: task.id.toString(), // dragId 값을 문자열로 변환하여 할당
  }));

  setConvertedTasks({
    ...convertedTasks,
    PROGRESS_IN: inStatusTasks,
  });

  const completeStatusTasks = tasks.PROGRESS_COMPLETE.map((task) => ({
    ...task,
    dragId: task.id.toString(), // dragId 값을 문자열로 변환하여 할당
  }));

  setConvertedTasks({
    ...convertedTasks,
    PROGRESS_COMPLETE: completeStatusTasks,
  });

  return convertedTasks;
}
