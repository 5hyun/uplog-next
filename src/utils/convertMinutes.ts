// ** 타이머 분:초 구하는 함수 */
export function convertMinutes(seconds: number): string {
  const minutes = Math.floor(seconds / 60);
  const remainingSeconds = seconds % 60;
  const formattedSeconds = remainingSeconds < 10 ? `0${remainingSeconds}` : `${remainingSeconds}`;

  if (minutes === 0) {
    return `00:${formattedSeconds}`;
  }

  return `${minutes}:${formattedSeconds}`;
}
