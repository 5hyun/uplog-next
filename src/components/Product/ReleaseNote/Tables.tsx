import React, { useCallback, useState } from 'react';
import { Tbody, Td, Tr } from '@chakra-ui/react';
import { GoKebabHorizontal } from 'react-icons/go';
import { AiOutlinePlus } from 'react-icons/ai';
import { useNavigate, useParams } from 'react-router-dom';
import { useRecoilState, useRecoilValue } from 'recoil';
import { typeBgColors } from '@/recoil/Product/ReleaseNote.ts';
import { eachProductProjects } from '@/recoil/Project/atom.ts';
import { editorChangeLog } from '@/recoil/Common/atom.ts';
import { UseQueryResult } from 'react-query';
import { ChangeLogData, Product, ProductInfo } from '@/typings/product.ts';
import { SaveUserInfo } from '@/typings/member.ts';

interface Props {
  isClickKebab: boolean;
  setIsClickKebab: React.Dispatch<React.SetStateAction<boolean>>;
  onClickKebab: (e: React.MouseEvent<HTMLButtonElement>) => void;
  onClickComplete: (modalType: 'add' | 'complete') => void;
  setTempVersion: React.Dispatch<React.SetStateAction<string>>;
  setNowProjectId: React.Dispatch<React.SetStateAction<number>>;
  eachProjectQueryResults: UseQueryResult<
    ChangeLogData[] | 'fail getChangeLogEachProject',
    unknown
  >[];
}

export default function Tables({
  isClickKebab,
  setIsClickKebab,
  onClickKebab,
  onClickComplete,
  setTempVersion,
  setNowProjectId,
  eachProjectQueryResults,
}: Props) {
  const [editChangeLog, setEditChangeLog] = useRecoilState(editorChangeLog);
  const reverseEachProjectQueryResults = [...eachProjectQueryResults].reverse();

  const [projects, setProjects] = useRecoilState(eachProductProjects);

  const navigate = useNavigate();
  const { product, project } = useParams();
  const nowProduct: ProductInfo = JSON.parse(sessionStorage.getItem('nowProduct')!);

  // 마우스 올렸을 때, kebab 버튼 나타남
  const [isHovering, setIsHovering] = useState(false);

  const handleMouseEnter = useCallback((status: string) => {
    if (status === 'PROGRESS_IN') setIsHovering(true);
  }, []);

  const handleMouseLeave = useCallback((status: string) => {
    if (status === 'PROGRESS_IN') {
      setIsHovering(false);
      setIsClickKebab(false);
    }
  }, []);

  return (
    <Tbody>
      {[...projects].reverse()?.map((version, index) => {
        return (
          <Tr
            key={index}
            color={version.projectStatus === 'PROGRESS_IN' ? 'var(--gray-dark)' : ''}
            position={'relative'}
            onMouseEnter={() => handleMouseEnter(version.projectStatus)}
            onMouseLeave={() => handleMouseLeave(version.projectStatus)}
          >
            <Td
              borderTop={'1px solid var(--gray-table)'}
              borderBottom={'1px solid var(--gray-table)'}
              padding={'1.6rem 0'}
              textAlign={'center'}
              cursor={'pointer'}
              onClick={() => {
                const url = encodeURI(`/workspace/${product}/${version.id}`);
                sessionStorage.setItem('nowProject', JSON.stringify(version));
                {
                  nowProduct.powerType === 'CLIENT'
                    ? navigate(`${url}/menu/결과물`)
                    : navigate(url);
                }
              }}
            >
              {version.version}
            </Td>
            <Td
              borderTop={'1px solid var(--gray-table)'}
              borderBottom={'1px solid var(--gray-table)'}
              textAlign={'center'}
            >
              {version.projectStatus === 'PROGRESS_IN' ? '진행 중' : version.endDate}
            </Td>
            <Td
              borderTop={'1px solid var(--gray-table)'}
              borderBottom={'1px solid var(--gray-table)'}
              marginY={'auto'}
              fontSize={'1.1rem'}
            >
              {eachProjectQueryResults[index]['data'] !== 'fail getChangeLogEachProject' &&
              Array.isArray(eachProjectQueryResults[index]['data']) ? (
                (
                  [
                    ...JSON.parse(JSON.stringify(eachProjectQueryResults?.[index]?.['data'])),
                  ] as ChangeLogData[]
                )?.map((content, idx) => {
                  return (
                    <div
                      key={idx}
                      className={`${
                        eachProjectQueryResults[index]['data']!.length - 1 !== idx && 'mb-5'
                      } flex items-center`}
                    >
                      <span
                        className={`${content.issueStatus === 'FEATURE' && 'bg-type-FEATURE'}
                       ${content.issueStatus === 'NEW' && 'bg-type-NEW'} 
                       ${content.issueStatus === 'CHANGED' && 'bg-type-CHANGED'}
                       ${content.issueStatus === 'FIXED' && 'bg-type-FIXED'}
                       ${
                         content.issueStatus === 'DEPRECATED' && 'bg-type-DEPRECATED'
                       } text-[#292723] mr-4 p-2 rounded-[0.31rem]`}
                      >
                        {content.issueStatus}
                      </span>
                      {content.title}
                    </div>
                  );
                })
              ) : (
                <button
                  className={
                    'flex items-center text-[1.2rem] text-gray-light font-bold cursor-pointer'
                  }
                  onClick={() => {
                    navigate(`/workspace/${product}/newchange`);
                    sessionStorage.setItem('nowProject', JSON.stringify(version));
                    setEditChangeLog('');
                  }}
                >
                  {version.projectStatus === 'PROGRESS_IN' && (
                    <>
                      <AiOutlinePlus className={'text-[1.6rem] mr-3'} /> 변경이력 추가
                    </>
                  )}
                </button>
              )}
            </Td>

            {/* Buttons for the "going" status */}
            {nowProduct?.powerType === 'MASTER' &&
              version.projectStatus === 'PROGRESS_IN' &&
              isHovering && (
                <Td
                  position={'absolute'}
                  top={'-0.5rem'}
                  right={'-1rem'}
                  backgroundColor={'inherit'}
                  border={'none'}
                >
                  <button onClick={onClickKebab}>
                    <GoKebabHorizontal className={'fill-gray-dark text-2xl cursor-pointer'} />
                  </button>
                  {isClickKebab && (
                    <div
                      className={
                        'absolute top-10 right-6 flex flex-col w-[7.5rem] h-20 border-solid border-[0.5px] border-gray-spring rounded text-xs shadow-release bg-white z-50'
                      }
                    >
                      <button
                        className={'h-1/2 hover:bg-orange-light-sideBar'}
                        onClick={() => {
                          navigate(`/workspace/${product}/newchange`);
                          sessionStorage.setItem('nowProject', JSON.stringify(version));
                          setEditChangeLog('');
                        }}
                      >
                        변경이력 추가
                      </button>
                      <button
                        className={'h-1/2 hover:bg-orange-light-sideBar'}
                        onClick={() => {
                          onClickComplete('complete');
                          setTempVersion(version.version);
                          setNowProjectId(version.id);
                        }}
                      >
                        완료
                      </button>
                    </div>
                  )}
                </Td>
              )}
          </Tr>
        );
      })}
    </Tbody>
  );
}
