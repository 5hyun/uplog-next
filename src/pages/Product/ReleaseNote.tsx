import React, { useCallback, useEffect, useLayoutEffect, useState } from 'react';
import { ProductInfo } from '@/typings/product.ts';
import { Table, TableContainer, Th, Thead, Tr, useDisclosure } from '@chakra-ui/react';
import { AiOutlinePlus } from 'react-icons/ai';
import Tables from '@/components/Product/ReleaseNote/Tables.tsx';
import ProjectModal from '@/components/Product/ReleaseNote/ProjectModal.tsx';
import { useLocation, useNavigate, useParams } from 'react-router-dom';
import { useGetAllProduct } from '@/components/Product/hooks/useGetAllProduct.ts';
import { useQueries, useQuery } from 'react-query';
import { getAllProductProjects, getChangeLogEachProject } from '@/api/Project/Version.ts';
import { useRecoilState } from 'recoil';
import { eachProductProjects } from '@/recoil/Project/atom.ts';
import { FaUserCircle } from 'react-icons/fa';
import { message } from '@/recoil/Common/atom.ts';
export default function ReleaseNote() {
  const [projects, setProjects] = useRecoilState(eachProductProjects);
  const nowProduct: ProductInfo = JSON.parse(sessionStorage.getItem('nowProduct')!);
  const { product } = useParams();

  const queryResults = useQueries([
    {
      queryKey: ['getAllProductProjects', nowProduct?.productId],
      queryFn: () => getAllProductProjects(nowProduct?.productId),
      staleTime: 300000, // 5분
      cacheTime: 600000, // 10분
      refetchOnMount: false, // 마운트(리렌더링)될 때 데이터를 다시 가져오지 않음
      refetchOnWindowFocus: false, // 브라우저를 포커싱했을때 데이터를 가져오지 않음
      refetchOnReconnect: false, // 네트워크가 다시 연결되었을때 다시 가져오지 않음
      enabled: !!nowProduct?.productId,
    },
  ]);

  const eachProjectQueryResults = useQueries(
    projects.map((project) => ({
      queryKey: ['getChangeLog', project.id],
      queryFn: () => getChangeLogEachProject(project.id),
      enabled: !!projects,
    }))
  );

  const navigate = useNavigate();
  const [productList, refetch, isFetching] = useGetAllProduct()!;
  const [messageInfo, setMessageInfo] = useRecoilState(message);

  const { isOpen, onOpen, onClose } = useDisclosure();
  // 모달 프로젝트 추가로 띄울건지, 프로젝트 완료로 띄울건지
  const [isAdd, setIsAdd] = useState(true);
  const [isClickKebab, setIsClickKebab] = useState(false);
  // 프로젝트 완료에서 임시 이름 저장하는 state
  const [tempVersion, setTempVersion] = useState('');
  const [nowProjectId, setNowProjectId] = useState(-1);

  /** 케밥 버튼 모달*/
  const onClickKebab = useCallback((e: React.MouseEvent<HTMLButtonElement>) => {
    e.stopPropagation();

    setIsClickKebab((prev) => !prev);
  }, []);

  const onCloseKebab = useCallback(() => {
    setIsClickKebab(false);
  }, []);

  // 프로젝트 모달
  const onClickProjectModal = useCallback((modalType: 'add' | 'complete') => {
    setIsAdd(modalType === 'add');
    onOpen();
  }, []);

  useEffect(() => {
    const projectList = queryResults[0].data;
    if (projectList && typeof projectList !== 'string') {
      const temp = JSON.parse(JSON.stringify(projectList));
      setProjects([...temp]);
    }
  }, [queryResults[0].data]);

  // 로그인 페이지에서 넘어왔을 경우
  useLayoutEffect(() => {
    if (isFetching) return;

    if (product === '-1' && productList?.length > 0) {
      sessionStorage.setItem('nowProduct', JSON.stringify(productList[0]));
      navigate(`/workspace/${productList[0].productId}`);
    } else if (product === '-1' && productList.length === 0) {
      navigate('/');
    }
  }, [product, productList, isFetching]);

  return (
    <section className={'w-full min-w-[50em] py-32 px-14 xl:px-56'} onClick={onCloseKebab}>
      <div className={'min-w-[30em] flex justify-between mb-4'}>
        <span className={'flex items-center'}>
          {nowProduct?.productImage ? (
            <img src={nowProduct?.productImage} alt={'제품 사진'} className={'w-12 h-12 mr-4'} />
          ) : (
            <FaUserCircle className={'w-12 h-12 mr-4 ml-2 text-3xl'} />
          )}
          <span className={'text-[2.4rem] font-bold'}>{nowProduct?.productName}</span>
        </span>
        <button
          className={'mr-2 text-gray-dark font-bold underline self-end'}
          onClick={() => {
            if (projects?.[0].projectStatus === 'PROGRESS_IN') {
              setMessageInfo({ type: 'warning', content: '현재 진행 중인 프로젝트가 있습니다.' });
              return;
            }

            setTempVersion('');
            onClickProjectModal('add');
          }}
        >
          프로젝트 추가하기
        </button>
      </div>

      <div>
        <TableContainer overflow={'hidden'} minHeight={'12rem'}>
          <Table fontSize={'1.2rem'} fontWeight={700}>
            <Thead>
              <Tr>
                <Th
                  borderTop={'1px solid var(--gray-table)'}
                  borderBottom={'1px solid var(--gray-table)'}
                  fontSize={'1.2rem'}
                  padding={'1.6rem 0'}
                  textAlign={'center'}
                  width={'20%'}
                  color={'var(--black)'}
                >
                  버전
                </Th>
                <Th
                  borderTop={'1px solid var(--gray-table)'}
                  borderBottom={'1px solid var(--gray-table)'}
                  fontSize={'1.2rem'}
                  padding={'1.6rem 0'}
                  textAlign={'center'}
                  width={'20%'}
                  color={'var(--black)'}
                >
                  날짜
                </Th>
                <Th
                  borderTop={'1px solid var(--gray-table)'}
                  borderBottom={'1px solid var(--gray-table)'}
                  fontSize={'1.2rem'}
                  padding={'1.6rem 0'}
                  textAlign={'center'}
                  width={'60%'}
                  color={'var(--black)'}
                >
                  변경이력
                </Th>
              </Tr>
            </Thead>
            {projects?.length > 0 && (
              <Tables
                isClickKebab={isClickKebab}
                setIsClickKebab={setIsClickKebab}
                onClickKebab={onClickKebab}
                onClickComplete={onClickProjectModal}
                setTempVersion={setTempVersion}
                setNowProjectId={setNowProjectId}
                eachProjectQueryResults={eachProjectQueryResults.reverse()}
              />
            )}
          </Table>
          {projects?.length === 0 && (
            <div
              className={'flex-row-center mt-6 text-gray-light text-2xl font-bold cursor-pointer'}
              onClick={onOpen}
            >
              <AiOutlinePlus className={'mr-4 text-3xl'} /> 프로젝트 시작하기
            </div>
          )}
        </TableContainer>
      </div>

      {/* 프로젝트 추가 완료 모달 */}
      <ProjectModal
        isOpen={isOpen}
        onClose={onClose}
        isAdd={isAdd}
        versionName={tempVersion}
        setMessageInfo={setMessageInfo}
        nowProjectId={nowProjectId}
      />
    </section>
  );
}
