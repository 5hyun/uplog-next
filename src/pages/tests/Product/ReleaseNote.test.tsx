import { describe, expect, test } from '@jest/globals';
import { fireEvent, getByText, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { withRouter } from '@/tests/utils';
import { Route } from 'react-router-dom';
import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';
import ProjectModal from '../../../components/Product/ReleaseNote/ProjectModal';
import ReleaseNote from '@/pages/Product/ReleaseNote';

describe('ReleaseNote', () => {
  test('ReleaseNote modal correctly', () => {
    const component = renderer.create(
      withRouter(<Route path={'/'} element={<ReleaseNote />} />, '/workspace/product')
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('ReleaseNote modal correctly', () => {
    const component = renderer.create(
      withRouter(<Route path={'/'} element={<ReleaseNote />} />, '/workspace/product')
    );

    expect(component.toJSON()).toMatchSnapshot();
  });
});
