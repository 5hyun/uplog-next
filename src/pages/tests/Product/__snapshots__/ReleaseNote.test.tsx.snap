// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReleaseNote ReleaseNote modal correctly 1`] = `null`;

exports[`ReleaseNote ReleaseNote modal correctly 2`] = `null`;
